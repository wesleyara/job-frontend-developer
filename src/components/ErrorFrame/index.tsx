import { MdErrorOutline } from "react-icons/md";

interface ErrorFrameProps {
  error: Error | null;
}

export const ErrorFrame = ({ error }: ErrorFrameProps) => {
  return (
    <section
      data-cy="error-frame"
      className="mt-5 w-full px-3 flex flex-col items-center gap-3"
    >
      <MdErrorOutline size={50} />
      <span className="text-center">{error?.message}</span>
    </section>
  );
};
