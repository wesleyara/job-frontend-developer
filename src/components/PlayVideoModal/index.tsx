import { Dialog, DialogPanel, DialogTitle } from "@headlessui/react";
import { YoutubeDetailsType } from "~/types";
import { Dispatch, SetStateAction } from "react";

interface PlayVideoModalProps {
  item: YoutubeDetailsType;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}

export const PlayVideoModal = ({
  item,
  isOpen,
  setIsOpen,
}: PlayVideoModalProps) => {
  return (
    <Dialog
      open={isOpen}
      onClose={() => setIsOpen(false)}
      className="relative z-50"
    >
      <div className="fixed inset-0 flex mx-auto items-center justify-center p-4">
        <DialogPanel className="space-y-4 border border-gray-300 bg-white p-12 max-w-[768px] w-[90%] text-center rounded-md relative">
          <button
            data-cy="close-youtube-modal"
            className="absolute top-2 right-3"
            onClick={() => setIsOpen(false)}
          >
            X
          </button>
          <DialogTitle className="text-lg">{item.snippet.title}</DialogTitle>

          <iframe
            data-cy="youtube-player"
            src={`https://www.youtube.com/embed/${item.id}`}
            title={item.snippet.title}
            className="inset-0 w-full h-[200px] sm:h-[400px]"
            allowFullScreen
          />
        </DialogPanel>
      </div>
    </Dialog>
  );
};
