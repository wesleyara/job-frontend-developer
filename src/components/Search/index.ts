import { SearchButton } from "./SearchButton";
import { SearchInput } from "./SearchInput";
import { SearchRoot } from "./SearchRoot";

export const Search = {
  Root: SearchRoot,
  Button: SearchButton,
  Input: SearchInput,
};
