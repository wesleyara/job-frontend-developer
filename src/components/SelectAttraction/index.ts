import { SelectAttractionButton } from "./SelectAttractionButton";
import { SelectAttractionContent } from "./SelectAttractionContent";
import { SelectAttractionRoot } from "./SelectAttractionRoot";

export const SelectAttraction = {
  Root: SelectAttractionRoot,
  Content: SelectAttractionContent,
  Button: SelectAttractionButton,
};
