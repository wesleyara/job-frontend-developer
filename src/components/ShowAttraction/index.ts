import { ShowAttractionBadges } from "./ShowAttractionBadges";
import { ShowAttractionInfo } from "./ShowAttractionInfo";
import { ShowAttractionRoot } from "./ShowAttractionRoot";

export const ShowAttraction = {
  Root: ShowAttractionRoot,
  Info: ShowAttractionInfo,
  Badges: ShowAttractionBadges,
};
