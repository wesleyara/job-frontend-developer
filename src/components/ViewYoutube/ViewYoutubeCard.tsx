"use client";

import { YoutubeDetailsType } from "~/types";
import Image from "next/image";
import { useState } from "react";
import { AiFillLike } from "react-icons/ai";
import { FaPlay, FaRegComment } from "react-icons/fa";
import { GrView } from "react-icons/gr";

interface ViewYoutubeCardProps {
  item: YoutubeDetailsType;
}

export const ViewYoutubeCard = ({ item }: ViewYoutubeCardProps) => {
  const [playHover, setPlayHover] = useState<boolean>(false);
  const [currentVideo, setCurrentVideo] = useState<string>("");

  const handleMouseEnter = (id: string) => {
    setCurrentVideo(id);
    setPlayHover(true);
  };

  const handleMouseLeave = () => {
    setCurrentVideo("");
    setPlayHover(false);
  };

  const showPlay = (id: string) => {
    return playHover && currentVideo === id;
  };

  const renderCounters = (count: string) => {
    const countNumber = Number(count);

    if (countNumber > 1000000000) {
      return `${(countNumber / 1000000000).toFixed(1)} bi`;
    }

    if (countNumber > 1000000) {
      return `${(countNumber / 1000000).toFixed(1)} mi`;
    }

    if (countNumber > 1000) {
      return `${(countNumber / 1000).toFixed(1)} mil`;
    }

    return countNumber;
  };

  return (
    <article
      key={item.id}
      className="flex flex-col gap-3 p-3 rounded-md border border-gray-300 bg-white w-full items-center"
    >
      <span
        className="relative cursor-pointer"
        onMouseEnter={() => handleMouseEnter(item.id)}
        onMouseLeave={() => handleMouseLeave()}
      >
        <FaPlay
          className={`play-tooltip ${showPlay(item.id) ? "show" : ""}`}
          size={40}
          color="#fff"
        />
        <Image
          src={item.snippet.thumbnails.medium.url}
          alt={item.snippet.title}
          width={320}
          height={180}
        />
      </span>
      <h3 className="text-lg font-bold">{item.snippet.title}</h3>
      <span className="flex gap-5">
        <span className="flex gap-1 items-center justify-center">
          <GrView /> {renderCounters(item.statistics.viewCount)}
        </span>
        <span className="flex gap-1 items-center justify-center">
          <AiFillLike /> {renderCounters(item.statistics.likeCount)}
        </span>
        <span className="flex gap-1 items-center justify-center">
          <FaRegComment /> {renderCounters(item.statistics.commentCount)}
        </span>
      </span>
      {item.snippet.description && (
        <span className="text-center">
          {item.snippet.description.slice(0, 100)}...
        </span>
      )}
    </article>
  );
};
