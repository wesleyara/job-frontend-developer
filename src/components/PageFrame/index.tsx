"use client";

import { useData } from "~/context/DataContext";
import { useState } from "react";
import { FaSearch } from "react-icons/fa";

import { ErrorFrame } from "../ErrorFrame";
import { Search } from "../Search";
import { SelectAttraction } from "../SelectAttraction";
import { ShowAttraction } from "../ShowAttraction";
import { ViewYoutube } from "../ViewYoutube";

export const PageFrame = () => {
  const {
    searchQuery,
    dataTicketSearch,
    dataYoutubeVideosDetails,
    selectedAttraction,
    isLoadingTicketSearch,
    isLoadingYoutubeVideos,
    isLoadingYoutubeVideosDetails,
    errorTicketSearch,
    errorYoutubeVideos,
    errorYoutubeVideosDetails,
    handleChange,
    handleSelectAttraction,
  } = useData();

  const [ratioSelected, setRatioSelected] = useState<string>("");
  const [seeMore, setSeeMore] = useState<boolean>(false);

  const enableToRender = dataTicketSearch && dataTicketSearch.length > 0;

  const render = () => {
    if (isLoadingTicketSearch) {
      return <p className="mt-10">Carregando...</p>;
    }

    if (errorTicketSearch || errorYoutubeVideos || errorYoutubeVideosDetails) {
      return <ErrorFrame error={errorTicketSearch} />;
    }

    if (selectedAttraction) {
      const imagesPerRatio: any = [];

      selectedAttraction.images.forEach(image => {
        const ratio = image.ratio;
        const find = imagesPerRatio.find((item: any) => item.ratio === ratio);

        if (find) {
          find.images.push(image);
        } else {
          imagesPerRatio.push({
            ratio,
            images: [image],
          });
        }
      });

      const selected = imagesPerRatio.find(
        (item: any) => item.ratio === ratioSelected,
      );

      return (
        <>
          <ShowAttraction.Root onClick={() => setSeeMore(prev => !prev)}>
            {!seeMore ? (
              <>
                <ShowAttraction.Info selectedAttraction={selectedAttraction} />
                <ShowAttraction.Badges
                  classifications={selectedAttraction.classifications}
                />
              </>
            ) : (
              <ShowAttraction.Carousel
                setRatioSelected={setRatioSelected}
                selected={selected}
                imagesPerRatio={imagesPerRatio}
                ratioSelected={ratioSelected}
              />
            )}
          </ShowAttraction.Root>

          <span className="flex mt-5">
            {isLoadingYoutubeVideos || isLoadingYoutubeVideosDetails ? (
              <p>Carregando...</p>
            ) : (
              !!dataYoutubeVideosDetails && (
                <ViewYoutube.Root>
                  {dataYoutubeVideosDetails?.map(item => (
                    <ViewYoutube.Card key={item.id} item={item} />
                  ))}
                </ViewYoutube.Root>
              )
            )}
          </span>
        </>
      );
    }

    if (enableToRender) {
      return (
        <div className="flex flex-col gap-3 mt-5 overflow-auto md:w-[600px] w-[90%]">
          {dataTicketSearch.map(item => (
            <SelectAttraction.Root key={item.id}>
              <SelectAttraction.Content item={item} />
              <SelectAttraction.Button
                onClick={() => handleSelectAttraction(item)}
              />
            </SelectAttraction.Root>
          ))}
        </div>
      );
    }
  };

  return (
    <main className="flex flex-col items-center my-24">
      <Search.Root>
        <Search.Input onChange={handleChange} value={searchQuery} />
        <Search.Button icon={FaSearch} />
      </Search.Root>

      {render()}
    </main>
  );
};
