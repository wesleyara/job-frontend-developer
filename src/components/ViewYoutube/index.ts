import { ViewYoutubeCard } from "./ViewYoutubeCard";
import { ViewYoutubeRoot } from "./ViewYoutubeRoot";

export const ViewYoutube = {
  Root: ViewYoutubeRoot,
  Card: ViewYoutubeCard,
};
