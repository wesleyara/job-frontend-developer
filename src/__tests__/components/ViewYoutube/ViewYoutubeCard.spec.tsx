import { render } from "@testing-library/react";
import { youtubeVideosDetailsMock } from "~/__tests__/mocks";
import { ViewYoutube } from "~/components/ViewYoutube";
import { beforeEach, describe, expect, it } from "vitest";

describe("ShowAttractionInfo unit test", () => {
  let sut: JSX.Element;

  beforeEach(() => {
    sut = <ViewYoutube.Card item={youtubeVideosDetailsMock} />;
  });

  it("should be defined", () => {
    expect(sut).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(sut);

    expect(container).toMatchSnapshot();
  });

  it("should have a span element", () => {
    const { container } = render(sut);

    expect(container.querySelector("span")).not.toBeNull();
  });

  it("should have a img element", () => {
    const { container } = render(sut);

    expect(container.querySelector("img")).not.toBeNull();
  });

  it("should have a icon element", () => {
    const { container } = render(sut);

    expect(container.querySelector("svg")).not.toBeNull();
  });

  it("should have a tailwind class name", () => {
    const { container } = render(sut);

    const element = container.querySelector(".flex");

    expect(element).not.toBeNull();
  });
});
