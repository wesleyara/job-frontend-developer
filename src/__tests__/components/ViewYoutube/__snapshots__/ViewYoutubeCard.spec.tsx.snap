// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShowAttractionInfo unit test > should render correctly 1`] = `
<div>
  <article
    class="flex flex-col gap-3 p-3 rounded-md border border-gray-300 bg-white w-full items-center"
  >
    <span
      class="relative cursor-pointer"
    >
      <svg
        class="play-tooltip "
        color="#fff"
        fill="currentColor"
        height="40"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(255, 255, 255);"
        viewBox="0 0 448 512"
        width="40"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
        />
      </svg>
      <img
        alt="Red Hot Chili Peppers - Californication (Official Music Video) [HD UPGRADE]"
        data-nimg="1"
        decoding="async"
        height="180"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fi.ytimg.com%2Fvi%2FYlUKcNNmywk%2Fmqdefault.jpg&w=640&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.ytimg.com%2Fvi%2FYlUKcNNmywk%2Fmqdefault.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ytimg.com%2Fvi%2FYlUKcNNmywk%2Fmqdefault.jpg&w=640&q=75 2x"
        style="color: transparent;"
        width="320"
      />
    </span>
    <h3
      class="text-lg font-bold"
    >
      Red Hot Chili Peppers - Californication (Official Music Video) [HD UPGRADE]
    </h3>
    <span
      class="flex gap-5"
    >
      <span
        class="flex gap-1 items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,21 C7,21 1,16 1,12 C1,8 7,3 12,3 C17,3 23,8 23,12 C23,16 17,21 12,21 Z M12,7 C9.23875,7 7,9.23875 7,12 C7,14.76125 9.23875,17 12,17 C14.76125,17 17,14.76125 17,12 C17,9.23875 14.76125,7 12,7 L12,7 Z"
            fill="none"
            stroke-width="2"
          />
        </svg>
         
        1.2 bi
      </span>
      <span
        class="flex gap-1 items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M885.9 533.7c16.8-22.2 26.1-49.4 26.1-77.7 0-44.9-25.1-87.4-65.5-111.1a67.67 67.67 0 0 0-34.3-9.3H572.4l6-122.9c1.4-29.7-9.1-57.9-29.5-79.4A106.62 106.62 0 0 0 471 99.9c-52 0-98 35-111.8 85.1l-85.9 311h-.3v428h472.3c9.2 0 18.2-1.8 26.5-5.4 47.6-20.3 78.3-66.8 78.3-118.4 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7-.2-12.6-2-25.1-5.6-37.1zM112 528v364c0 17.7 14.3 32 32 32h65V496h-65c-17.7 0-32 14.3-32 32z"
          />
        </svg>
         
        5.1 mi
      </span>
      <span
        class="flex gap-1 items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 32C114.6 32 0 125.1 0 240c0 47.6 19.9 91.2 52.9 126.3C38 405.7 7 439.1 6.5 439.5c-6.6 7-8.4 17.2-4.6 26S14.4 480 24 480c61.5 0 110-25.7 139.1-46.3C192 442.8 223.2 448 256 448c141.4 0 256-93.1 256-208S397.4 32 256 32zm0 368c-26.7 0-53.1-4.1-78.4-12.1l-22.7-7.2-19.5 13.8c-14.3 10.1-33.9 21.4-57.5 29 7.3-12.1 14.4-25.7 19.9-40.2l10.6-28.1-20.6-21.8C69.7 314.1 48 282.2 48 240c0-88.2 93.3-160 208-160s208 71.8 208 160-93.3 160-208 160z"
          />
        </svg>
         
        131.4 mil
      </span>
    </span>
    <span
      class="text-center"
    >
      Watch the official music video for Californication by Red Hot Chili Peppers from the album Californi
      ...
    </span>
  </article>
</div>
`;
